### PR TITLE
Multiple Levels in Activities , Part 1

### LINKED ISSUE #137: Multiple levels of difficulty for existing games ( Part I )
## Description
As a [USER],
I need to have various levels in the games,
so that I can enjoy playing them even more.

### Update [Required]
- [ ] Picture puzzle game enhancement
- [ ] Ranking logic changes 
- [ ] Picture puzzle game sanity test
- [ ] Do's and don'ts game enhancement
- [ ] Do’s and Don’ts Game ranking logic
- [ ] Odd one out game (including penalty points reduction)
- [ ] Ranking logic changes
- [ ] Odd one out game sanity test
- [ ] Odd one out points reduction and ranking logic changes unit testing (Front End)
- [ ] Odd one out points reduction and ranking logic changes unit testing (Back End)

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentors.

## Estimation
31 - 37 hours

### PR DESCRIPTION
**Pull request by lunayach**
### Description
Added multiple levels in the following activities:

- Picture Puzzle Component
- Odd One Out Component
- Do's and Don'ts  Component
- Memory Game Component

Fixes #137

### Type of Change:

- Code
- User Interface
- Outreach
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Wrote new tests in Karma and Mocha. Previous Karma and Mocha Tests passed.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged 
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

#### UI of the Navigate Level Component (middle Right) which would be used by all activities having multiple levels. (Only one activity is shown in GIF)
final9

### CODE REVIEW

**Comment by Buddhiprabha** [comment]
Can we increase the cell width of the memory game? The cells look cluttered as levels go up. Also the idea was to allow the user have the option to go to the next level or to the next next game. @Raxerz @shreyans29 @paavininanda

File: config/passport.js [hunk 1]
```diff
@@ -76,7 +77,21 @@ module.exports = function(passport, models) {
                         .catch(function(err) {
                             return done(err);
                         });
-                })
+                    levelProgress.findOrCreate({where: {
+                        user_id: user.id
+                    }, defaults: {
+                        user_id: user.id,
+                    }})
+                        .spread(() => {
```
Comment (shreyans29) [inline config/passport.js]: Identation here..

File: data/codes.json [hunk 2]
```diff
@@ -158,10 +158,30 @@
...
+            "code": "PCE038",
+            "message": "Something went wrong while updating level-progress data",
+            "reason": "Internal Server Error"
+      },
+        "PCE039": {
+            "code": "PCE039",
+            "message": "Could not update user's activity level",
+            "reason": "Request body was empty"
+      }
     },
```
Comment (shreyans29) [inline data/codes.json]: I think this line can be removed.

File: src/app/ActivityConfig.ts [hunk 3]
```diff
@@ -1,47 +1,59 @@
...
+    levels: 3
   },
   {
     name: 'Moskito Asasinato',
-    url: 'unlocked-stage'
+    url: 'unlocked-stage',
+    levels: 3
   },
 ];
+
```
Comment (shreyans29) [inline src/app/ActivityConfig.ts]: Nice design :)

File: src/app/introduction/activity-2/activity-2.component.html [hunk 4]
```diff
@@ -9,5 +9,5 @@
       <p>{{language?.descriptions.bodyContent4 +' '+ email}}</p>
     </div>
   </div>
-  <button-nav [completed]="completed" nexturl="/introduction/activity-3" prevurl="/introduction/activity-1"></button-nav>
+  <button-nav [completed]="completed" nexturl="/introduction/activity-3/1" prevurl="/introduction/activity-1"></button-nav>
```
Comment (shreyans29) [inline src/app/introduction/activity-2/activity-2.component.html]: No need to change previous url?

File: src/app/level-navigate/level-navigate.component.scss [hunk 5]
```diff
@@ -0,0 +1,18 @@
+@import "../styles/main";
+
+#level-navigate {
+  display: inline-block;
+  position: absolute;
+  top: 50%;
+  right: 0;
```
Comment (shreyans29) [inline src/app/level-navigate/level-navigate.component.scss]: I think we have a variable for 0.. $zero I guess.

File: src/app/level-navigate/level-navigate.component.scss [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+button {
+  border-radius: $border-radius;
+  font-size: $font-size-small;
+}
+
+.active-level {
```
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/level-navigate/level-navigate.component.scss]: 4 for css 2 for HTML

File: src/app/level-navigate/level-navigate.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+
+describe('LevelNavigateComponent', () => {
+  let component: LevelNavigateComponent;
+  let fixture: ComponentFixture<LevelNavigateComponent>;
+  const mockLevel = 1;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [ LevelNavigateComponent ],
```
Comment (shreyans29) [inline src/app/level-navigate/level-navigate.component.spec.ts]: Space not needed.

File: src/app/level-navigate/level-navigate.component.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/level-navigate/level-navigate.component.scss]: Okay, Got it. :+1:

File: data/codes.json (hunk 2 above)
Comment (lunayach) [reply data/codes.json]: But like all the Error Codes have followed this pattern until now. So, this may not look consistent.

File: src/app/introduction/activity-2/activity-2.component.html (hunk 4 above)
Comment (lunayach) [reply src/app/introduction/activity-2/activity-2.component.html]: @shreyans29  Not for now, because multiple levels for those activities have not been done in this PR.

File: src/app/level-navigate/level-navigate.component.scss (hunk 5 above)
Comment (lunayach) [reply src/app/level-navigate/level-navigate.component.scss]: Currently, we did not have this variable. So added it.

**Comment by lunayach** [comment]
Addressed the comments.

File: src/app/level-navigate/level-navigate.component.spec.ts (hunk 7 above)
Comment (lunayach) [reply src/app/level-navigate/level-navigate.component.spec.ts]: Okay.

File: config/passport.js (hunk 1 above)
Comment (shreyans29) [reply config/passport.js]: . spread can be started from previous line itself..

File: data/codes.json (hunk 2 above)
Comment (shreyans29) [reply data/codes.json]: no no, I meant the empty line in middle

Comment (shreyans29) [reply data/codes.json]: line 186

File: src/app/introduction/activity-2/activity-2.component.html (hunk 4 above)
Comment (shreyans29) [reply src/app/introduction/activity-2/activity-2.component.html]: got it

File: src/app/performance-display/performance-display.component.ts [hunk 8]
```diff
@@ -33,8 +34,11 @@ export class PerformanceDisplayComponent implements OnInit {
     this.activityName = this.SuccessObject.currentActivityName;
     this.nextActivityURL = this.SuccessObject.nextActivityURL;
     this.nextActivity = this.SuccessObject.nextActivity;
-    document.getElementById('message').click();
-    document.getElementById('message').focus();
+    this.currActivityURL = this.SuccessObject.currActivityURL;
+    this.nextActivityLevels = this.SuccessObject.nextActivityLevels;
+    for (let i = 2; i <= this.SuccessObject.levels; i++) {
```
Comment (shreyans29) [inline src/app/performance-display/performance-display.component.ts]: add a semicolon there..

File: src/app/meds-n-labels/activity-2/activity-2.component.ts [hunk 9]
```diff
@@ -36,26 +39,34 @@ export class MemoryGameComponent implements OnInit {
...
+                'rapid-diagnostic-test-1.png',
                 'symptoms-1.png',
                 'symptoms-1.png',
-                'enlarged-spleen-1.png',
-                'area-1.png',
                 'virus-1.png',
-                'rapid-diagnostic-test-1.png',
                 'virus-1.png',
             ];
+    public levelConfig = [{cardsCount: 8, rowCount: 4, columnCount: 2},
```
Comment (shreyans29) [inline src/app/meds-n-labels/activity-2/activity-2.component.ts]: move things to new line after [

File: config/passport.js [hunk 10]
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
...
+                            const response = {email: user.email, name: user.name};
+                            return done(null, response);
+                        })
+                        .catch(function(err) {
+                            return done(err);
+                        });
+                }
+
```
Comment (paavininanda) [inline config/passport.js]: Extra line here also, I think this can be removed.

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 11]
```diff
@@ -9,6 +9,9 @@ import { BadgeService } from '../../services/BadgeService/badge.service';
 import swal from 'sweetalert2';
 import { PerformanceDisplayService } from '../../services/performance-display.service';
 
+import { ActivatedRoute } from '@angular/router';
+
```
Comment (paavininanda) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: Extra space here also

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 12]
```diff
@@ -91,11 +106,10 @@ export class OddOneOutComponent implements OnInit {
         } else {
             this._sharedData.customErrorAlert(this.alerts.activityFailMsg, this.alerts.activityFailTitle);
         }
-        this._obs = observableInterval(1000).pipe(
+        this._obs = observableInterval(10).pipe(
```
Comment (paavininanda) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: Why did you change this?

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 13]
```diff
@@ -32,9 +37,14 @@ export class OddOneOutComponent implements OnInit {
     public completed = false;
     public alerts: any;
     public solutions = '';
+    public level: number;
+    public levelConfig = [4, 6, 8]; // no of questions depending upon the level
```
Comment (paavininanda) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: Number of questions is okay, but don't we have separate questions for each level ?

File: src/app/introduction/activity-2/activity-2.component.html (hunk 4 above)
Comment (Raxerz) [reply src/app/introduction/activity-2/activity-2.component.html]: @lunayach please make note of such changes that need to be added in the next PR.
It is very easy to miss them.

File: test/test.js [hunk 14]
```diff
@@ -291,6 +291,37 @@ describe('Testing APIs', function() {
...
+                    .set('x-access-token', `'${token}'`)
+                    .send(body)
+                    .expect(200)
+                    .end(function(err, res) {
+                        res.body.level.should.equal(1);
+                        done();
+                    });
+            });
+
+
```
Comment (shreyans29) [inline test/test.js]: Do these tests cover all possible scenarios.

File: database/models/levelProgress.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        moskitoAsasinato: {
+            type: Sequelize.INTEGER,
+            defaultValue: 0
+        },
+        user_id: {
+            type: Sequelize.INTEGER,
+            unique: true
+        }
+    }, {
+        tableName: 'levelProgress'
```
Comment (Raxerz) [inline database/models/levelProgress.js]: table name change with underscore separation

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 16]
```diff
@@ -364,11 +374,13 @@ export class PicturePuzzleComponent implements OnInit {
...
         this.activityComplete = true;
         if (!this.completed) {
-          const badgeNumber = 1;
-          const currStage = 2;
-          this._performanceService.openDialog(currStage);
-          this._badgeService.updateBadgeNumber(badgeNumber).subscribe(res => res);
+          this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
+          if (this.level === 1) {
+            const badgeNumber = 1;
+            const currStage = 2;
```
Comment (Raxerz) [inline src/app/introduction/activity-3/activity-3.component.ts]: constant naming convention

File: src/app/level-navigate/level-navigate.component.html [hunk 17]
```diff
@@ -0,0 +1,9 @@
+<div id="level-navigate" *ngIf="levelsURL.length > 1">
+  <div *ngFor="let levelURL of levelsURL; index as i;">
+    <a [href]="levelURL" [ngClass]="{'active-level': (i+1) === userCurrentLevel }">
```
Comment (Raxerz) [inline src/app/level-navigate/level-navigate.component.html]: can the comparison operation be moved to the ts file?

File: src/app/malaria-101/activity-2/activity-2.component.ts [hunk 18]
```diff
@@ -16,6 +17,8 @@ export class DragdropComponent implements OnInit {
     public dropCheckDo: String;
     public dropCheckDont: String;
     language: any;
+    private score = 0;
+    private pointsPerCorrectAnswer = 10;
```
Comment (Raxerz) [inline src/app/malaria-101/activity-2/activity-2.component.ts]: constant naming convention

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 19]
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
     // "#video" is the name of the template variable in the video element
     @ViewChild('video') video: any;
 
-    private PUZZLE_DIFFICULTY = 4;
+    private PUZZLE_DIFFICULTY: number;
+    private difficultyList = [2, 4, 6];
```
Comment (shreyans29) [inline src/app/introduction/activity-3/activity-3.component.ts]: will this be changed ? if not make it constant. Same with levels.

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 20]
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
     // "#video" is the name of the template variable in the video element
     @ViewChild('video') video: any;
 
-    private PUZZLE_DIFFICULTY = 4;
+    private PUZZLE_DIFFICULTY: number;
```
Comment (shreyans29) [inline src/app/introduction/activity-3/activity-3.component.ts]: This is not a read only variable. Please use camelCase.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 19 above)
Comment (paavininanda) [reply src/app/introduction/activity-3/activity-3.component.ts]: And also include this level list in some common file. Do the same for all activities

File: src/app/malaria-101/activity-2/activity-2.component.ts [hunk 21]
```diff
@@ -25,6 +28,10 @@ export class DragdropComponent implements OnInit {
     public position= 'col-md-10 col-md-offset-2 introbody';
     public completed = false;
     public alerts: any;
+    public levelConfig = [5, 7, 10];
```
Comment (paavininanda) [inline src/app/malaria-101/activity-2/activity-2.component.ts]: This can be stored in some common file.

**Comment by lunayach** [comment]
@paavininanda  Actually, I have done changes locally. Would push them now.

File: config/passport.js (hunk 1 above)
Comment (lunayach) [reply config/passport.js]: Done.

File: data/codes.json (hunk 2 above)
Comment (lunayach) [reply data/codes.json]: Done.

File: src/app/performance-display/performance-display.component.ts (hunk 8 above)
Comment (lunayach) [reply src/app/performance-display/performance-display.component.ts]: Done.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts (hunk 9 above)
Comment (lunayach) [reply src/app/meds-n-labels/activity-2/activity-2.component.ts]: Done.

File: config/passport.js (hunk 10 above)
Comment (lunayach) [reply config/passport.js]: Done.

File: src/app/malaria-101/activity-3/activity-3.component.ts (hunk 11 above)
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Done.

File: src/app/malaria-101/activity-3/activity-3.component.ts (hunk 12 above)
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Here, 1000 was the time delay between displaying the next question, when the user has answered the previous one. So, this being not a smooth experience, I reduced the delay to 10ms.

File: src/app/malaria-101/activity-3/activity-3.component.ts (hunk 13 above)
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Yup. Added.

File: test/test.js (hunk 14 above)
Comment (lunayach) [reply test/test.js]: Done.

File: database/models/levelProgress.js (hunk 15 above)
Comment (lunayach) [reply database/models/levelProgress.js]: Done.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 16 above)
Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Done in the previous PR.

File: src/app/level-navigate/level-navigate.component.html (hunk 17 above)
Comment (lunayach) [reply src/app/level-navigate/level-navigate.component.html]: To bind a different class to specific HTML element, we would need to do a comparison inline, this way. There may not be a .ts way for this.

File: src/app/malaria-101/activity-2/activity-2.component.ts (hunk 18 above)
Comment (lunayach) [reply src/app/malaria-101/activity-2/activity-2.component.ts]: Done in the previous PR.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 19 above)
Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Done. for making **readonly**

For storing in some common file, may not work here. Because, the way levels are increased in different activities is different. That is sometimes, it is just changing no. of fragments like in this case and then somtimes, it involves fetching a different property from a JSON file like objects marked under 'level1', 'level2' etc. Ex. for different set of questions for different levels.
So having a **readonly** activity-specific declaration at the top, may not be a bad idea.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 20 above)
Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Done.

**Comment by lunayach** [comment]
Addressed comments.

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 22]
```diff
@@ -19,7 +20,10 @@ export class PicturePuzzleComponent implements OnInit {
     // "#video" is the name of the template variable in the video element
     @ViewChild('video') video: any;
 
-    private PUZZLE_DIFFICULTY = 4;
+    private puzzleDifficulty: number;
+    private readonly DIFFICULTY_LIST = [4, 6, 8];
+    private levels = ['level1', 'level2', 'level3'];
```
Comment (Raxerz) [inline src/app/introduction/activity-3/activity-3.component.ts]: can be a constant?

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 23]
```diff
@@ -57,9 +61,11 @@ export class PicturePuzzleComponent implements OnInit {
     public completed = false;
     public alerts: any;
     public userData: any;
+    public currStage = 2;
```
Comment (Raxerz) [inline src/app/introduction/activity-3/activity-3.component.ts]: constant?

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 24]
```diff
@@ -56,9 +66,15 @@ export class OddOneOutComponent implements OnInit {
...
         });
+        this._route.params.subscribe( params => {
+          this.noOfQuestions = this.levelConfig[ params.level - 1];
+          this.level =  parseInt(params.level, 10);
+        });
         this._dashboardService.getJSONData('quiz.json').subscribe(response => {
             this._data = JSON.parse(response.data);
-            this.shuffle(this._data.quizlist);
+            this._data = this._data[this.levels[this.level - 1]];
+            console.log(JSON.stringify(this._data));
```
Comment (Raxerz) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: remove the log statement

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 22 above)
Comment (shreyans29) [reply src/app/introduction/activity-3/activity-3.component.ts]: yep that and naming convention

Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Yes. This would change when the previous PR is merged.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 23 above)
Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Done in the previous PR.

File: src/app/malaria-101/activity-3/activity-3.component.ts (hunk 24 above)
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Removed.

**Comment by lunayach** [comment]
Addressed comments.

**Comment by Buddhiprabha** [comment]
@lunayach why is @shreyans29 's review dismissed?

**Comment by lunayach** [comment]
@Buddhiprabha even if we make one line change as compared to the approved version. Review gets dismissed. I didn't dismiss it. GitHub automatically does it.

**Comment by paavininanda** [comment]
@lunayach Could you resolve the conflicts ?

File: routes/api.js [hunk 25]
```diff
@@ -158,6 +159,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
...
+router.post('/user/activity/level', authenticationHelpers.isAuthOrRedirect, (req, res) => {
+    if (!req.user.email) {
+        return res.status(400).json({error: errorCode.PCE021.message, code: errorCode.PCE021.code});
+    }
+    localUser.find({where: {
+        email: req.user.email
+    }}, {raw: true})
+        .then(data => {
+            if (!data) {
+                return res.status(200).json({info: errorCode.PCE022.message, code: errorCode.PCE022.code});
```
Comment (shreyans29) [inline routes/api.js]: error code..

File: routes/api.js [hunk 26]
```diff
@@ -158,6 +159,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
...
+            }
+            levelProgress.find({where: {
+                user_id: data.id
+            }}, {raw: true})
+                .then(activityData => {
+                    if (!activityData) {
+                        return res.status(200).json({info: errorCode.PCE036.message, code: errorCode.PCE036.code});
```
Comment (shreyans29) [inline routes/api.js]: error code

File: routes/api.js (hunk 25 above)
Comment (lunayach) [reply routes/api.js]: @shreyans29 error code seems correct to me?

File: routes/api.js (hunk 26 above)
Comment (lunayach) [reply routes/api.js]: For this one as well.

File: routes/api.js (hunk 25 above)
Comment (paavininanda) [reply routes/api.js]: He means to say that res.status should not be 200 since its an error. It should be 402.

Comment (lunayach) [reply routes/api.js]: Okay! Got it.

File: routes/api.js [hunk 27]
```diff
@@ -159,6 +160,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
...
+router.post('/user/activity/levelNumber', authenticationHelpers.isAuthOrRedirect, (req, res) => {
+    if (!req.user.email) {
+        return res.status(400).json({error: errorCode.PCE021.message, code: errorCode.PCE021.code});
+    }
+    localUser.find({where: {
+        email: req.user.email
+    }}, {raw: true})
+        .then(data => {
+            if (!data) {
+                return res.status(402).json({info: errorCode.PCE022.message, code: errorCode.PCE022.code});
```
Comment (Raxerz) [inline routes/api.js]: 402 error code?

File: routes/api.js [hunk 28]
```diff
@@ -159,6 +160,66 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
...
+            }
+            levelProgress.find({where: {
+                user_id: data.id
+            }}, {raw: true})
+                .then(activityData => {
+                    if (!activityData) {
+                        return res.status(402).json({info: errorCode.PCE036.message, code: errorCode.PCE036.code});
```
Comment (Raxerz) [inline routes/api.js]: 402 is again used? I don't think its right

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 29]
```diff
@@ -369,10 +377,15 @@ export class PicturePuzzleComponent implements OnInit {
         this.initPuzzle();
         this.activityComplete = true;
         if (!this.completed) {
-          this._performanceService.openDialog(this.CURR_STAGE);
-          this._badgeService.updateBadgeNumber(this.BADGE_NUMBER).subscribe(res => res);
-          this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: 'level1'})
+          this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: this.LEVELS[this.level - 1]});
+
+          this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
```
Comment (Raxerz) [inline src/app/introduction/activity-3/activity-3.component.ts]: can 'oddOneOut' be made a constant?

File: routes/api.js (hunk 27 above)
Comment (shreyans29) [reply routes/api.js]: I think it should be 400

File: routes/api.js (hunk 28 above)
Comment (shreyans29) [reply routes/api.js]: yes it should be 400 as in the whole app.

File: app.js [hunk 30]
```diff
@@ -29,7 +29,7 @@ winston.add(
 const models = require('./database/models');
 
 // Sync Database
-models.sequelize.sync().then(function() {
+models.sequelize.sync({force: true}).then(function() {
```
Comment (shreyans29) [inline app.js]: if this is what you referred to that day, please remove it.  It should not be the the branch. You can keep using it for dev purposes.

File: routes/api.js (hunk 28 above)
Comment (Raxerz) [reply routes/api.js]: @shreyans29 I think it should be 200 only, the same is being discussed in the next PR.

Comment (lunayach) [reply routes/api.js]: @shreyans29, @Raxerz may be correct here. Because **200** actually means that **if** block worked properly. So, for a Developer, this would mean that exceptions are handled correctly and hence success.

File: app.js (hunk 30 above)
Comment (lunayach) [reply app.js]: Okay! Corrected it.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts [hunk 31]
```diff
@@ -166,15 +180,19 @@ export class MemoryGameComponent implements OnInit {
...
                 this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
                 this._status = {stage: 3, activity: 2};
                 this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
                 this.activityComplete = true;
                 if (!this.completed) {
-                  this._badgeService.updateBadgeNumber(this.BADGE_NUMBER).subscribe(res => res);
-                  this._performanceService.openDialog(this.CURR_STAGE); }
-                  this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: 'level1'});
+                  this._leaderBoardService.updateLeaderBoard({activity: this.ACTIVITY, level: this.LEVELS[this.levelNumber - 1]});
+                  this._dashboardService.updateActivityLevel({activity: 'memoryGame', level: this.levelNumber}).subscribe(() => {});
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/meds-n-labels/activity-2/activity-2.component.ts]: Done.

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 32]
```diff
@@ -147,8 +157,11 @@ export class OddOneOutComponent implements OnInit {
               html: this.solutions,
               type: 'info',
             }).then( res => { if (!this.completed) {
-              this._badgeService.updateBadgeNumber(this.BADGE_NUMBER).subscribe(response => response);
-              this._performanceService.openDialog(this.CURR_STAGE);
+              this._dashboardService.updateActivityLevel({activity: 'oddOneOut', level: this.level}).subscribe(() => {});
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Done.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 29 above)
Comment (lunayach) [reply src/app/introduction/activity-3/activity-3.component.ts]: Done.

**Comment by Raxerz** [comment]
@lunayach the activities that have been changed are not working for some reason.
screenshot from 2018-08-01 23 07 49
screenshot from 2018-08-01 23 06 53

**Comment by lunayach** [comment]
@Raxerz please check now. Actually, there was just a mistake in naming variables. :sweat_smile:

**Comment by Raxerz** [comment]
@lunayach It works fine now but there seems to be issues. Please check the following as well.
I have selected the wrong options but I still get a success notification
screenshot from 2018-08-03 01 10 25

The options did not appear at certain times, this was inconsistent but usually happened when I answered the questions too quickly.
screenshot from 2018-08-03 01 10 59

I got the success message but the level navigator arrow to move to the next level didn't appear
screenshot from 2018-08-03 01 11 27

Didn't we decide on having level 3 as level 1 here?
screenshot from 2018-08-03 01 13 52

**Comment by Raxerz** [comment]
@lunayach It works fine now but there seems to be issues. Please check the following as well.
I have selected the wrong options but I still get a success notification
screenshot from 2018-08-03 01 10 25

The options did not appear at certain times, this was inconsistent but usually happened when I answered the questions too quickly.
screenshot from 2018-08-03 01 10 59

I got the success message but the level navigator arrow to move to the next level didn't appear
screenshot from 2018-08-03 01 11 27

The image is cutting off the border here and make the changes related to the size as suggested by buddhi.
screenshot from 2018-08-03 01 12 53

Didn't we decide on having level 3 as level 1 here, I don't see that change made, have you made it in the second PR?
screenshot from 2018-08-03 01 13 52

**Comment by shreyans29** [comment]
@lunayach please address the issues @Raxerz  has mentioned asap.. there are lots of PRs still pending..

**Comment by lunayach** [comment]
@Raxerz done the changes.
For **1**, It comes when you use all available sentences. We allow him to go to the next activity even if his all options are not correct. We do keep a track of no. of correct answers and accordingly, he gets a dynamic score in LeaderBoard.

For **2**, There was just a problem in renaming a variable. Fixed it.
For **3**, It's a desirable behaviour. Please click **next** to complete next set and completing that would complete the activity. Anyway, this behaviour is changed in the next PR.
For **4, 5** Fixed border and added those images.

**Comment by Raxerz** [comment]
@lunayach for **1** how will the user know if he was right or not? This is very misleading.

File: src/app/introduction/activity-3/activity-3.component.ts (hunk 29 above)
Comment (Raxerz) [reply src/app/introduction/activity-3/activity-3.component.ts]: I don't see this change made

File: src/app/meds-n-labels/activity-2/activity-2.component.spec.ts [hunk 33]
```diff
@@ -28,6 +30,7 @@ describe('MemoryGameComponent', () => {
   let apiService: APIService;
   let sharedService: SharedDataService;
   let languageService: LanguageService;
+  const mockLevel = 1;
```
Comment (Raxerz) [inline src/app/meds-n-labels/activity-2/activity-2.component.spec.ts]: constant naming convention

File: src/app/meds-n-labels/activity-2/activity-2.component.ts [hunk 34]
```diff
@@ -77,6 +95,14 @@ export class MemoryGameComponent implements OnInit {
             this.language = response.pcprepkit.stages.medsNLabels.memoryGame;
             this.alerts = response.pcprepkit.common.alerts;
         });
+        this._route.params.subscribe( params => {
+        this.currentLevel = this.LEVEL_CONFIG[ params.level - 1];
+        this.levelNumber =  parseInt(params.level, 10);
+          if (this.levelNumber === 3) {
```
Comment (Raxerz) [inline src/app/meds-n-labels/activity-2/activity-2.component.ts]: make the 3 a constant

**Comment by Raxerz** [comment]
@paavininanda please check and see if this PR is fine so that it can be merged.
